File: library/src/main/java/com/licrafter/tagview/TagAdapter.java
package com.licrafter.tagview;

import android.database.DataSetObservable;
import android.database.DataSetObserver;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import com.licrafter.tagview.views.ITagView;

/**
 * Created by lijx on 2017/5/17.
 */

public abstract class TagAdapter {

    private static final boolean DEBUG = false;
    private static final String TAG = "TagAdapter";
    private DataSetObservable mObservable = new DataSetObservable();

    /**
     * Return the number of ITagView.
     */
    public abstract int getCount();

    /**
     * Return the ITagView associated with a specified position.
     */
    public abstract ITagView getItem(int position);

    public ITagView instantiateItem(ViewGroup container, int position) {
        long itemId = getItemId(position);
        ITagView tagView = getItem(position);
        if (DEBUG) Log.v(TAG, "Adding tagView #" + itemId + ": TagView=" + tagView);
        container.addView((View) tagView);
        return tagView;
    }

    public boolean isViewFromObject(View view, Object object) {
        return object.equals(view);
    }

    public void destroyItem(ViewGroup container, int position, ITagView object) {

    }

    public int getItemId(int position) {
        return position;
    }

    public void notifyDataSetChanged() {
        mObservable.notifyChanged();
    }

    public void registerDataSetObserver(DataSetObserver observer) {
        mObservable.registerObserver(observer);
    }

    public void unregisterDataSetObserver(DataSetObserver observer) {
        mObservable.unregisterObserver(observer);
    }

}

File: library/src/main/kotlin/com/licrafter/tagview/TagAdapter.kt
package com.licrafter.tagview

import android.database.DataSetObservable
import android.database.DataSetObserver
import android.util.Log
import android.view.View
import android.view.ViewGroup
import com.licrafter.tagview.views.ITagView

/**
 * Created by lijx on 2018/5/7.
 * Gmail: shellljx@gmail.com
 */
abstract class TagAdapter {

    private var mObservable = DataSetObservable()

    /**
     * Return the number of ITagView.
     */
    abstract fun getCount(): Int

    abstract fun getItem(position: Int): ITagView

    fun instantiateItem(container: ViewGroup, position: Int): ITagView {
        val itemId = getItemId(position).toLong()
        val tagView = getItem(position)
        if (DEBUG) Log.v(TAG, "Adding tagView #$itemId: TagView=$tagView")
        container.addView(tagView as View)
        return tagView
    }

    fun isViewFromObject(view: View, obj: Any): Boolean {
        return obj == view
    }

    fun destroyItem(container: ViewGroup, position: Int, tagView: ITagView) {

    }

    fun getItemId(position: Int): Int {
        return position
    }

    fun notifyDataSetChanged() {
        mObservable.notifyChanged()
    }

    fun registerDataSetObserver(observer: DataSetObserver) {
        mObservable.registerObserver(observer)
    }

    fun unregisterDataSetObserver(observer: DataSetObserver) {
        mObservable.unregisterObserver(observer)
    }

    companion object {
        const val DEBUG = false
        const val TAG = "TagAdapter"
    }
}